assets/aml-benchmark/components/src/dataset_preprocessor/dataset_preprocessor.py

```diff
@@ -0,0 +1,136 @@
+        self,
+        input_dataset: str = None,
+        template: str = None,
+        user_preprocessor: str = None,
+        encoder_config: str = None,
+        output_dataset: str = None
+    ):
+        """Dataset Preprocessor Class.
+
+        :param input_dataset: Path to the directory to load the dataset.
+        :param template: A json dictionary where key is the name of the column enclosed in " " and associated \
+            dict value is presented using jinja template logic which will be used to extract the \
+            respective value from the dataset.
+        :param user_preprocessor: Path to the custom preprocessor python script provided by user.
+        :param encoder_config: Dictionary in json format that contains idtolabel or labeltoid to encode. This is not \
+            applicable to custom scripts.
+        :param output_dataset: Path to the dump the processed .jsonl file.
+        """
+        self.input_dataset = input_dataset
+        self.template = template
+        self.user_preprocessor = user_preprocessor
+        self.encoder_config = encoder_config
+        self.output_dataset = output_dataset
+        self.validate()
+
+    def validate(self):
+        """Validate the parameters."""
+        if self.input_dataset is None:
+            mssg = (
+                "Path to load the dataset is not provided."
+            )
+            raise BenchmarkValidationException._with_error(
+                AzureMLError.create(BenchmarkValidationError, error_details=mssg)
+            )
+        if self.template is None and self.user_preprocessor is None:
+            mssg = (
+                "Please provide the input to apply preprocessing logic either via template input or \
+                script_path."
+            )
+            raise BenchmarkValidationException._with_error(
+                AzureMLError.create(BenchmarkValidationError, error_details=mssg)
+            )
+        if self.user_preprocessor and not self.user_preprocessor.endswith('.py'):
+            mssg = (
+                "Please provide python script containing your custom preprocessor logic."
+            )
+            raise BenchmarkValidationException._with_error(
+                AzureMLError.create(BenchmarkValidationError, error_details=mssg)
+            )
+
+    def add_json_filter(self, template):
+        """Add tojson filter in the template."""
+        patt = r'({{.*?}})'
+        pat = re.compile(patt)
+        matches = re.findall(pat, template)
+        for m in range(len(matches)):
+            logger.info(f"{template}, {matches[m]}")
+            s1 = '{{'+matches[m].lstrip('{{').rstrip('}}')+'|tojson}}'
+            template = template.replace(matches[m], s1)
+        logger.info(f"Final template:{template}")
+        return template
+
+    def prep_using_template(self):
+        """Preprocessor run using template."""
+        data = read_jsonl_files([self.input_dataset])
+        template = json.dumps(self.template)

```

Comment:
Doing `dumps` then `loads` seems a bit strange. Why do we need this?

Reply:
With P1 datasets that we have, I have come across some uncleaned datasets containing utf-8, which became a problem where json.loads did not work but dumping and then loading it worked.